File: javascript/list-ops/list-ops.js
export class List {
  constructor(values = []) {
    this.values = [...values];
  }

  append(list) {
    const newList = new List([...this.values]);
    
    for (let i = 0; i < list.values.length; i++) {
      newList.values.push(list.values[i]);
    }
    
    return newList;
  }

  concat(listOfLists) {
    const newList = new List([...this.values]);
    
    for (let i = 0; i < listOfLists.values.length; i++) {
      const currentList = listOfLists.values[i];
      
      for (let j = 0; j < currentList.values.length; j++) {
        newList.values.push(currentList.values[j]);
      }
    }
    
    return newList;
  }

  filter(predicate) {
    const newList = new List();
    
    for (let i = 0; i < this.values.length; i++) {
      if (predicate(this.values[i])) {
        newList.values.push(this.values[i]);
      }
    }
    
    return newList;
  }

  length() {
    let count = 0;
    
    for (let i = 0; i < this.values.length; i++) {
      count++;
    }
    
    return count;
  }

  map(transform) {
    const newList = new List();
    
    for (let i = 0; i < this.values.length; i++) {
      newList.values.push(transform(this.values[i]));
    }
    
    return newList;
  }

  foldl(reducer, initial) {
    let accumulator = initial;
    
    for (let i = 0; i < this.values.length; i++) {
      accumulator = reducer(accumulator, this.values[i]);
    }
    
    return accumulator;
  }

  foldr(reducer, initial) {
    let accumulator = initial;
    
    for (let i = this.values.length - 1; i >= 0; i--) {
      accumulator = reducer(accumulator, this.values[i]);
    }
    
    return accumulator;
  }

  reverse() {
    const newList = new List();
    
    for (let i = this.values.length - 1; i >= 0; i--) {
      newList.values.push(this.values[i]);
    }
    
    return newList;
  }
}

File: python/list-ops/list_ops.py
def append(list1, list2):
    result = []
    for item in list1:
        result = result + [item]
    for item in list2:
        result = result + [item]
    return result


def concat(lists):
    result = []
    for list_item in lists:
        for item in list_item:
            result = result + [item]
    return result


def filter(function, list):
    result = []
    for item in list:
        if function(item):
            result = result + [item]
    return result


def length(list):
    count = 0
    for _ in list:
        count += 1
    return count


def map(function, list):
    result = []
    for item in list:
        result = result + [function(item)]
    return result


def foldl(function, list, initial):
    result = initial
    for item in list:
        result = function(result, item)
    return result


def foldr(function, list, initial):
    result = initial
    # Use our own reverse function to process the list from right to left
    reversed_list = reverse(list)
    for item in reversed_list:
        result = function(result, item)
    return result


def reverse(list):
    result = []
    # Manual implementation without using range
    i = length(list)
    while i > 0:
        i -= 1
        result = result + [list[i]]
    return result
